text/3374-cargo-target-features.md

```diff
@@ -0,0 +1,212 @@
+- Feature Name: `cargo_target_features`
+- Start Date: 2023-01-20
+- RFC PR: [rust-lang/rfcs#3374](https://github.com/rust-lang/rfcs/pull/3374)
+- Tracking Issue: [rust-lang/cargo#0000](https://github.com/rust-lang/cargo/issues/0000)
+
+# Summary
+[summary]: #summary
+
+This adds a new `enable-features` field to [Cargo Targets](https://doc.rust-lang.org/cargo/reference/cargo-targets.html#configuring-a-target) which forces a [feature](https://doc.rust-lang.org/cargo/reference/features.html) to be enabled if the target is being built.
+
+# Motivation
+[motivation]: #motivation
+
+There are several situations where a non-library target must have a particular feature enabled for it to work correctly.
+Although it is possible to manually enable these features on the command-line, this can make it awkward to use.
+This RFC adds a mechanism to make it easier to work with these targets.
+
+Some example use cases are:
+
+1. Binaries that need additional dependencies like command-line processing libraries, or logging functionality.
+2. Examples illustrating how to use a library with a specific feature enabled.
+3. Benchmarks which require a separate benchmarking library, but you don't want to pay the cost of building that library when not running benchmarks.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+The `enable-features` field can be added to a target in `Cargo.toml` to specify features or dependencies that will be automatically enabled whenever the target is being built.
+For example:
+
+```toml
+[package]
+name = "myproject"
+version = "0.1.0"
+edition = "2021"
+
+[[example]]
+name = "fetch-http"
+enable-features = ["networking"]
+
+[[bin]]
+name = "myproject"
+enable-features = ["dep:clap"]
+
+[dependencies]
+clap = { version = "4.0.26", optional = true }
+
+[features]
+networking = []
+```
+
+When using `myproject` as a library dependency, by default the `networking` feature and `clap` will not be enabled.
+When building the binary as in `cargo install myproject` or `cargo build`, the `clap` dependency will automatically be enabled allowing the binary to be built correctly.
+Similarly, `cargo build --example fetch-http` will enable the `networking` feature so that the example can be built.
```

Comment:
> automatically enabled whenever the target is being built\
> ...
> When building the binary as in `cargo install myproject` or `cargo build`, the `clap` dependency will automatically be enabled allowing the binary to be built correctly.
> ...
> Cargo will need to determine which features are being force-enabled by the targets selected on the command-line. 
> Unfortunately this selection of targets is quite complex. This may require some potentially invasive refactoring, as Cargo currently chooses the targets to select after resolution is done (including a separate phase for implicit binaries used for integration tests).

I feel like an important part of this RFC is non-obvious unless a person is very familiar with these terms and fully thinks through the implications of those terms.

Going into this RFC, my care about was
- `cargo run --example foo` would activate the needed features so I could run it

However, this RFC also introduces
- `cargo build` will build all default targets, activating their `enable-features`
- `cargo test` will build and test all default targets, activating their `enable-features`
- `cargo check --all-targets` will build and test all  targets, activating their `enable-features`

This is in contrast to  `required-features` which would have skipped targets.

My primary concern with this is it will make it impossible to `check` and `test` a project without something like [cargo-hack](https://crates.io/crates/cargo-hack).  In nearly all of my projects, CI runs
- `cargo test --workspace --no-default features`
- `cargo test --workspace`
- `cargo test --workspace --all-features`

The assumption is this covers enough feature combinations to feel confident that it will build.  There are combinations of features that might break but the probability is low enough to not justify extra CI time.

**With `enable-features`, these three test runs could become equivalent.**  The only way to verify in CI what I had originally intended would be to select the exact targets for the desired feature set which would be unmanageable without external tooling.

From watching cargo's issue tracker and other forums, I feel like feature unification is one of the biggest hurdles to features in cargo today and I feel that this design compounds that problem to being unmanageable.

Under the alternatives is listed:

> Only the situation where required-features generates an error could be changed to implicitly enable the missing features. This would likely make required-features less annoying to work with, but doesn't help for use cases like running cargo test where you have specific tests or examples that you want to be automatically included (where required-features simply makes them silently excluded).

As I said, this is what I expected going into this RFC.

For `cargo test` or `cargo build --examples`, having it skips incompatible targets seems fine.    When I want to indiscriminately enable all, that is what `--all-features` is for.  Granted, it can be annoying when you have debug features.  **Maybe another alternative is a `--required-features` flag that activates all required features for the selected targets?**

The one use case I can think of that wouldn't be served by this is `cargo install pulldown-cmark` ([manifest](https://github.com/raphlinus/pulldown-cmark/blob/master/Cargo.toml)).  Since no target is selected, `required-features` wouldn't activate but then there wouldn't be anything to install.  I suspect these are minor enough of cases that with improved error messages (https://github.com/rust-lang/rfcs/pull/3374#discussion_r1084069560), enabling of features instead of error, and `--required-features`, things would be sufficiently improved into a "good enough" state.

Reply:
> The one use case I can think of that wouldn't be served by this is cargo install pulldown-cmark ([manifest](https://github.com/raphlinus/pulldown-cmark/blob/master/Cargo.toml?rgh-link-date=2023-06-20T20%3A02%3A04Z)). Since no target is selected, required-features wouldn't activate but then there wouldn't be anything to install.

What's stopping us from default enabling `--required-features` on `install` command?

Reply:
I had considered including that but vaguely remember a conversation with concerns for a similar idea ("just make `cargo install` do the right thing").

One problem is there are two different workflows
- Bin-only packages where `required-features` would only be used in truly optional bins
- Primarily lib packages with optional bins at which point we'd need `required-features` for the most basic of `bin`s.

This also makes `cargo install` behave differently than nearly any cargo command (`cargo metadata` is an exception) which makes it less predictable.  In particular, this furthers the split `cargo build` and `cargo install`s behavior (handling of lockfiles being a point of contention). 

That isn't to say I'm against it but that we'd need to do weigh it carefully and maybe find ways to help with these problems.

Reply:
Those are very valid and strong concerns. Part of the drawbacks mentions `--no-default-features`, but I hadn't considered that just testing a library without other binaries enabling features you don't want could be problematic. My original thinking was that if that's what you want, `cargo test --lib --no-default-features` would be the command to run, but I understand how that's probably not realistic for most people to run.

`--no-default-features` could disable all `enable-features` clauses, though that seems worryingly complicated to explain, and fragile for testing.

> For cargo test or cargo build --examples, having it skips incompatible targets seems fine.

IIUC, this is suggesting to use `required-features` instead? That doesn't seem to address the use cases that I have in mind. For example, if a package has a library and a binary, and the binary has command-line processing dependencies, and the integration tests need to test the binary. If those binaries used `required-features`, then the tests would fail (or possibly just not run). That's not the expectation I have for a bare `cargo test` which by default should test the majority of things in the package.

Reply:
> IIUC, this is suggesting to use required-features instead? That doesn't seem to address the use cases that I have in mind. For example, if a package has a library and a binary, and the binary has command-line processing dependencies, and the integration tests need to test the binary. If those binaries used required-features, then the tests would fail (or possibly just not run). That's not the expectation I have for a bare cargo test which by default should test the majority of things in the package.

This doesn't seem prohibitive.  The tests could have `cfg`s for the relevant features needed for them to run, just like tests calling into parts of an API that are feature dependent.  Alternatively, you could declare specific `test` crates to have `required-features` that match the `bin`.  In place of individual features, you can define a `cli` feature (imo this is what *should* be done or else you are having people rely on implementation details).  I don't think this is enough to justify this feature.

Reply:
> That's not the expectation I have for a bare `cargo test` which by default should test the majority of things in the package.

I have not found this to be true in feature-heavy crates. I've taken two approaches in the past: feature-gate tests appropriately and note in the readme that you need to use `--features <stuff you touched>`, or add an internal feature and `compile_error!("please test with --all-features");` the tests if it's not active.

Reply:
Perhaps my comment wasn't clear, but for me `required-features` isn't a solution. For example, if I ran `cargo test` in cargo's repo, and that didn't run the testsuite (because the binary requires clap), I would be very frustrated. 

> I have not found this to be true up till now in feature-heavy crates.

This isn't intended for just feature-heavy situations. The primary use case and motivation for this is target-specific dependencies, where you might not have any features.

Reply:
And FWIW, it may be the solution is to instead lean more heavily on workspaces, and making them easier to use. That is what I heavily lean towards as an alternative to this, particularly since the amount of complexity ended up being far greater than I was expecting. I'm not sure what might be the solution there, but it would at least be helpful for all workspace users if workspaces were easier to use.

Reply:
> This isn't intended for just feature-heavy situations. The primary use case and motivation for this is target-specific dependencies, where you might not have any features.

I think this is one of the problems with the RFC.  It is taking a very specific use case / workflow (bin/lib crates existing in the same package) and making a general feature for it.  For example, test, and bench targets will more likely run into this problem due to being in feature-heavy crates (like [clap](https://github.com/clap-rs/clap/blob/master/Cargo.toml#L119)) for which this solution is a major step back yet is being advertised as solving that problem.

> And FWIW, it may be the solution is to instead lean more heavily on workspaces, and making them easier to use. T

imo potential steps forward are:
- Improve the `cargo install` error
- Improve the experience with `required-features`
  - Add a `--required-features` flag
  - For `--bin`, `--example`, and `--test` flags, enable the required-features
- Improve the workspace experience:
  - [private crates](https://internals.rust-lang.org/t/private-nested-cargo-packages/19042) (helps with workspaces generally, does not help with bin/lib splits in specific)
  - [exploring "recommended bins"](https://internals.rust-lang.org/t/better-ux-for-cargo-install-command-where-the-command-is-provided-by-another-crate/18215)
  - [publish multiple packages at once](https://github.com/rust-lang/cargo/issues/1169)
  - ...

Reply:
For the cargo-install use case, another solution is "bin dependencies" or dependencies that activate for bins.